core/src/main/scala/org/apache/spark/sql/sources/SparkOptimizations.scala

```diff
@@ -0,0 +1,399 @@
+        )
+
+    def unapply(tables: (LogicalPlan, LogicalPlan)):
+    Option[ReturnType] = {
+      val (left, right) = tables
+
+      /** now doing a one-to-one mapping of lefttable and its indexes with
+        * right table and its indexes. Following example explains the combination
+        * generator.
+        *
+        * val l = Seq(1, 2, 3)
+        * val r = Seq(4, 5)
+        * l.zip(Seq.fill(l.length)(r)).flatMap {
+        * case (leftElement, rightList) => rightList.flatMap { e =>
+        * Seq((leftElement, e))
+        * }
+        * }.foreach(println)
+        * will output :
+        * (1, 4)
+        * (1, 5)
+        * (2, 4)
+        * (2, 5)
+        * (3, 4)
+        * (3, 5)
+        *
+        */
+      val leftRightEntityMapping = for {
+        (leftTable, leftIndexes) <- fetchIndexes(left)
+        (rightTable, rightIndexes) <- fetchIndexes(right)
+        leftSeq = Seq(leftTable) ++ leftIndexes
+        rightSeq = Seq(rightTable) ++ rightIndexes
+      } yield {
+        leftSeq.zip(Seq.fill(leftSeq.length)(rightSeq)) flatMap {
+          case (leftElement, rightList) => rightList.flatMap(e => Seq((leftElement, e)))
+        }
+      }
+
+      // right now not expecting multiple tables in left & right hand side.
+      //      assert(leftRightEntityMapping.size <= 1)
+
+      val mappings = leftRightEntityMapping.flatMap { mappedElements =>
+        val (leftTable, rightTable) = mappedElements(0) // first pairing is always (table, table)
+        for {
+
+          (leftPlan, rightPlan) <- mappedElements
+
+          leftRelation = unwrapBaseColumnRelation(leftPlan) if leftRelation.nonEmpty
+          rightRelation = unwrapBaseColumnRelation(rightPlan) if rightRelation.nonEmpty
+
+          if isColocated(leftRelation.get, rightRelation.get)
+        } yield {
+
+          val leftReplacement = leftTable match {
+            case _: LogicalRelation => (leftTable, leftPlan)
+            case subquery@SubqueryAlias(alias, _) =>
+              (subquery, SubqueryAlias(alias, leftPlan))
+          }
+
+          val rightReplacement = rightTable match {
+            case _: LogicalRelation => (rightTable, rightPlan)
+            case subquery@SubqueryAlias(alias, _) =>
+              (subquery, SubqueryAlias(alias, rightPlan))
+          }
+
+          ((leftRelation.get, rightRelation.get), leftReplacement, rightReplacement)
+        }
+      }
+
+      if (mappings.nonEmpty) {
+        Some(mappings.unzip3)
+      } else {
+        None
+      }
+    }
+
+  }
+
+  private def unwrapBaseColumnRelation(
+      plan: LogicalPlan): Option[BaseColumnFormatRelation] = plan collectFirst {
+    case LogicalRelation(relation: BaseColumnFormatRelation, _, _) =>
+      relation
+    case SubqueryAlias(alias, LogicalRelation(relation: BaseColumnFormatRelation, _, _)) =>
+      relation
+  }
+
+  private def isColocated(left: BaseColumnFormatRelation, right: BaseColumnFormatRelation) = {
```

Comment:
not very clear about the last portion. Did you mean, create table ... partition_by `spark expr` ? Assuming two tables created over the api using same partitioner can take advantage of colocated join, this guy will simply think tables are non-colocated. If there are indexes defined by our way and those indexes on each table are declared as colocated, then yes, this will be replaced with indexes on both side. Isn't that we want ?